### PR TITLE
fix(Blade): add `size` prop to Text component and update tokens

### PR DESCRIPTION
**Pull request by kamleshchandnani**
This PR updates the typography tokens scale for mobile devices to create better visual hierarchy which we received as feedback from other teams as well.

It also adds a new `size` prop to `Text` component for `variant='body'`

### CODE REVIEW

File: packages/blade/src/components/Typography/Text/Text.tsx [hunk 1]
```diff
@@ -66,14 +69,17 @@ const getProps = <T extends { variant: TextVariant }>({
   };
 
   if (variant === 'body') {
-    props.fontSize = 100;
-    props.fontStyle = 'normal';
-    props.lineHeight = 'l';
```
Comment (chaitanyadeorukhkar) [inline packages/blade/src/components/Typography/Text/Text.tsx:71]: Why did you remove `fontStyl` & `lineHeight`?

File: packages/blade/src/components/Typography/Text/Text.tsx [hunk 2]
```diff
@@ -66,14 +69,17 @@ const getProps = <T extends { variant: TextVariant }>({
...
+    }
   } else if (variant === 'caption') {
     if (weight === 'bold') {
       throw new Error(`[Blade: Text]: weight cannot be 'bold' when variant is 'caption'`);
     }
-    props.fontSize = 25;
+    if (size === 'small') {
+      throw new Error(`[Blade: Text]: size cannot be 'small' when variant is 'caption'`);
+    }
+    props.fontSize = 50;
```
Comment (chaitanyadeorukhkar) [inline packages/blade/src/components/Typography/Text/Text.tsx:82]: Size for caption changed from 50 to 25?

File: packages/blade/src/components/Typography/Text/Text.tsx [hunk 3]
```diff
@@ -66,14 +69,17 @@ const getProps = <T extends { variant: TextVariant }>({
...
     props.fontStyle = 'italic';
     props.lineHeight = 's';
```
Comment (chaitanyadeorukhkar) [inline packages/blade/src/components/Typography/Text/Text.tsx:84]: Lineheight of `s` is good enough still since we increased the fontSize from 25 to 50?

Comment (kamleshchandnani) [reply packages/blade/src/components/Typography/Text/Text.tsx:84]: yep that was a typo on figma in the past we just rectfied it, @saurav12 to confirm if the lineheight is good enough even with `50`.

File: packages/blade/src/components/Typography/Text/Text.tsx (hunk 2 above)
Comment (kamleshchandnani) [reply packages/blade/src/components/Typography/Text/Text.tsx:82]: yes, past sins on figma

**Review by saurav12 — DISMISSED** [review]
This lgtm 👍. Will make the changes to Figma as discussed

File: packages/blade/src/components/Typography/Text/Text.tsx (hunk 3 above)
Comment (saurav12) [reply packages/blade/src/components/Typography/Text/Text.tsx:84]: Need to check it once. I will revert once I'm done with the changes on Figma. Have to check it visually.

**Review by chaitanyadeorukhkar — APPROVED** [review]

File: packages/blade/src/components/Button/BaseButton/buttonTokens.ts [hunk 4]
```diff
@@ -9,52 +8,29 @@ import type { BaseButtonProps } from './BaseButton';
...
-      xsmall: 's',
-      small: 's',
-      medium: 'm',
-      large: 'm',
-    },
+  lineHeights: {
+    xsmall: 's',
+    small: 's',
+    medium: 'l',
+    large: 'm',
```
Comment (divyanshu013) [inline packages/blade/src/components/Button/BaseButton/buttonTokens.ts:33]: Didn't quite understand the rationale for making medium `l` and large `m`

File: packages/blade/src/tokens/global/typography.ts [hunk 5]
```diff
@@ -19,12 +19,12 @@ export type FontSize = {
   25: number;
   /** desktop: 11(px/rem/pt)
    *
-   * mobile: 12(px/rem/pt)
+   * mobile: 11(px/rem/pt)
    */
```
Comment (anuraghazra) [inline packages/blade/src/tokens/global/typography.ts:23]: Yep this will fix the Checkbox onDesktop onMobile discrepancy. 💯

File: packages/blade/src/tokens/global/typography.ts [hunk 6]
```diff
@@ -191,33 +191,33 @@ const typography: TypographyWithPlatforms = {
...
         400: 20,
         500: 22,
-        600: 24,
-        700: 27,
+        600: 22,
+        700: 24,
         800: 29,
         900: 32,
-        1000: 35,
+        1000: 29,
```
[the remark this answers is not in the document]
Comment (anuraghazra) [reply packages/blade/src/tokens/global/typography.ts:205]: +1 wouldn't think break the hierarchy?

File: packages/blade/src/components/Button/BaseButton/buttonTokens.ts [hunk 7]
```diff
@@ -9,52 +8,29 @@ import type { BaseButtonProps } from './BaseButton';
...
-        NonNullable<BaseButtonProps['size']>,
-        keyof ThemeTokens['typography'][Key]['fonts']['size']
-      >;
-    };
-    lineHeights: Record<
-      NonNullable<BaseButtonProps['size']>,
-      keyof ThemeTokens['typography'][Key]['lineHeights']
-    >;
+  fonts: {
+    size: Record<NonNullable<BaseButtonProps['size']>, keyof Theme['typography']['fonts']['size']>;
```
Comment (anuraghazra) [inline packages/blade/src/components/Button/BaseButton/buttonTokens.ts:12]: **nit:** 
Just clean this up a bit. 

```ts
type ButtonSizes = NonNullable<BaseButtonProps['size']>;
// replace all repetitions
```

File: packages/blade/src/components/Typography/Text/__tests__/Text.native.test.tsx [hunk 8]
```diff
@@ -46,6 +46,35 @@ describe('<Text />', () => {
...
+    try {
+      const displayText = 'Displaying some text';
+      renderWithTheme(
+        // @ts-expect-error testing failure case when size='small' is passed with variant='caption'
+        <Text type="normal" variant="caption" truncateAfterLines={3} size="small">
+          {displayText}
+        </Text>,
+      );
+    } catch (error: unknown) {
+      if (error instanceof Error) {
```
Comment (anuraghazra) [inline packages/blade/src/components/Typography/Text/__tests__/Text.native.test.tsx:70]: Never noticed this. But ideally we should not call expect() conditionally. 

Say introduced a custom Error reported. `BladePropError()` this check will not pass.

**Comment by anuraghazra** [comment]
LGTM just resolve the conflicts. 
And what happened to [this comment?](https://github.com/razorpay/blade/pull/548#discussion_r920000521)

File: packages/blade/src/components/Button/BaseButton/buttonTokens.ts (hunk 7 above)
Comment (kamleshchandnani) [reply packages/blade/src/components/Button/BaseButton/buttonTokens.ts:10]: why so? we still do need to have this right?
```
fonts: {
    size: {
      xsmall: 75,
      small: 75,
      medium: 100,
      large: 200,
    },
  },
}
```

File: packages/blade/src/components/Button/BaseButton/buttonTokens.ts (hunk 4 above)
Comment (kamleshchandnani) [reply packages/blade/src/components/Button/BaseButton/buttonTokens.ts:31]: the left hand side are sizes of the button, not to be confused with the exact literal :D

File: packages/blade/src/components/Typography/Text/__tests__/Text.native.test.tsx (hunk 8 above)
Comment (kamleshchandnani) [reply packages/blade/src/components/Typography/Text/__tests__/Text.native.test.tsx:70]: if we don't add this check TS yells because it doesn't know what is error's type 😭

Comment (anuraghazra) [reply packages/blade/src/components/Typography/Text/__tests__/Text.native.test.tsx:70]: Yeah just set any.

File: packages/blade/src/components/Button/BaseButton/buttonTokens.ts (hunk 7 above)
Comment (anuraghazra) [reply packages/blade/src/components/Button/BaseButton/buttonTokens.ts:10]: It doesn't change anything. Just we are repeating it a lot. 


Screenshot 2022-07-26 at 4 06 45 PM

File: packages/blade/src/components/Button/BaseButton/buttonTokens.ts (hunk 4 above)
Comment (divyanshu013) [reply packages/blade/src/components/Button/BaseButton/buttonTokens.ts:31]: Does it imply having a large line height `l` for a `medium` button and so on?

File: packages/blade/src/tokens/global/typography.ts [hunk 9]
```diff
@@ -191,33 +191,33 @@ const typography: TypographyWithPlatforms = {
...
-      xl: 24,
-      '2xl': 28,
+      xl: 28,
+      '2xl': 24,
       '3xl': 28,
-      '4xl': 30,
+      '4xl': 32,
       '5xl': 30,
-      '6xl': 38,
+      '6xl': 40,
```
[the remark this answers is not in the document]
Comment (kamleshchandnani) [reply packages/blade/src/tokens/global/typography.ts:220]: yes that's something tragic since we have these tokens as aliases(common across native and web) and hence they are not very intuitive and not incrementing. not able to think of any other way 😢.

File: packages/blade/src/components/Button/BaseButton/buttonTokens.ts (hunk 4 above)
Comment (kamleshchandnani) [reply packages/blade/src/components/Button/BaseButton/buttonTokens.ts:31]: valid. @saurav12 can you check this once?

File: packages/blade/src/components/Typography/Text/Text.tsx (hunk 3 above)
Comment (kamleshchandnani) [reply packages/blade/src/components/Typography/Text/Text.tsx:85]: this is resolved

File: packages/blade/src/tokens/global/typography.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (kamleshchandnani) [reply packages/blade/src/tokens/global/typography.ts:205]: @saurav12 to check this.

File: packages/blade/src/tokens/global/typography.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (divyanshu013) [reply packages/blade/src/tokens/global/typography.ts:220]: Can we add some comment in the code to explain this? Because looking at tokens it seems very confusing that `xl` is `28` and `2xl` is smaller than it (`24`). Similarly for `4xl` and `5xl`

File: packages/blade/src/components/Button/BaseButton/buttonTokens.ts (hunk 4 above)
Comment (kamleshchandnani) [reply packages/blade/src/components/Button/BaseButton/buttonTokens.ts:31]: thanks for pointing this out.
Discussed with Saurav and we'll map

button - medium lineheight to `m`
button - large lineheight to `l`

**Review by divyanshu013 — DISMISSED** [review]
Looks good to me.
